File: Javascript/src/S1_SOLID/open-closed.ts
enum Color {
  Red,
  Green,
  Blue,
}
enum Size {
  Small,
  Medium,
  Large,
}

class Product {
  name: string;
  color: Color;
  size: Size;

  constructor(name: string, color: Color, size: Size) {
    this.name = name;
    this.color = color;
    this.size = size;
  }
}

// Broke Open-Closed principle
// class ProductFilter shouldn't Open for Modification (when we have a new Filter)
class ProductFilter {
  filterByColor(products: Product[], color: Color) {
    return products.filter((product) => product.color === color);
  }

  filterBySize(products: Product[], size: Size) {
    return products.filter((product) => product.size === size);
  }

  filterByColorAndSize(products: Product[], color: Color, size: Size) {
    return products.filter(
      (product) => product.color === color && product.size === size
    );
  }
}

const apple = new Product('Apple', Color.Green, Size.Small);
const tree = new Product('Tree', Color.Green, Size.Large);
const house = new Product('House', Color.Blue, Size.Large);
const products = [apple, tree, house];

const pf = new ProductFilter();
console.log('Green products (old):');
for (const product of pf.filterByColor(products, Color.Green)) {
  console.log(` - ${product.name} is green`);
}

//========================================================================
// Using Specification pattern to make sure Open-Closed principle
// Create 2 new Interface that are Open for Extension

interface Specification<T> {
  isSatisfied: (item: T) => boolean;
}
interface Filter<T> {
  filter: (items: T[], spec: Specification<T>) => T[];
}

class ColorSpecification implements Specification<Product> {
  color: Color;

  constructor(color: Color) {
    this.color = color;
  }

  isSatisfied(item: Product) {
    return item.color === this.color;
  }
}

class SizeSpecification implements Specification<Product> {
  size: Size;

  constructor(size: Size) {
    this.size = size;
  }

  isSatisfied(item: Product) {
    return item.size === this.size;
  }
}

// Combination
class AndSpecification<T> implements Specification<T> {
  specs: Specification<T>[];

  constructor(...specs: Specification<T>[]) {
    this.specs = specs;
  }

  isSatisfied(item: T) {
    return this.specs.every((spec) => spec.isSatisfied(item));
  }
}

class BetterFilter implements Filter<Product> {
  filter(items: Product[], spec: Specification<Product>) {
    return items.filter((item) => spec.isSatisfied(item));
  }
}

const bf = new BetterFilter();

console.log('Green product (new):');
for (const product of bf.filter(
  products,
  new ColorSpecification(Color.Green)
)) {
  console.log(` - ${product.name} is green`);
}

console.log('Large blue items:');
for (const product of bf.filter(
  products,
  new AndSpecification<Product>(
    new ColorSpecification(Color.Blue),
    new SizeSpecification(Size.Large)
  )
)) {
  console.log(` - ${product.name} is blue and large`);
}

File: CSharp/S1_SOLID/OpenClosed.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace S1_SOLID
{
    public static class OpenClosed
    {
        public static void Main()
        {
            var apple = new Product("Apple", Color.Green, Size.Small);
            var tree = new Product("Tree", Color.Green, Size.Large);
            var house = new Product("House", Color.Blue, Size.Large);
            Product[] products = { apple, tree, house };

            var pf = new ProductFilter();
            Console.WriteLine("Green products (old):");
            foreach (var p in pf.FilterByColor(products, Color.Green))
            {
                Console.WriteLine($" - {p.Name} is green");
            }

            var bf = new BetterFilter();
            Console.WriteLine("Green products (new):");
            foreach (var p in bf.Filter(products, new ColorSpecification(Color.Green)))
            {
                Console.WriteLine($" - {p.Name} is green");
            }
            Console.WriteLine("Large blue items:");
            foreach (var p in bf.Filter(products, new AndSpecification<Product>(new ColorSpecification(Color.Blue), new SizeSpecification(Size.Large))))
            {
                Console.WriteLine($" - {p.Name} is blue and large");
            }
        }
    }

    public enum Color { Red, Green, Blue }
    public enum Size { Small, Medium, Large }

    public class Product
    {
        public string Name { get; set; }
        public Color Color { get; set; }
        public Size Size { get; set; }

        public Product(string name, Color color, Size size)
        {
            Name = name ?? throw new ArgumentNullException(paramName: nameof(name));
            Color = color;
            Size = size;
        }
    }

    // Open-Closed Principle (OCP): Open for Extension, but Closed for Modification
    // Broke OCP, when we want to combine Color & Size, we Opened for Modification (FilterBySizeAndColor)
    public class ProductFilter
    {
        public IEnumerable<Product> FilterByColor(IEnumerable<Product> products, Color color)
        {
            foreach (Product product in products)
            {
                if (product.Color == color) yield return product;
            }
        }

        public IEnumerable<Product> FilterBySize(IEnumerable<Product> products, Size size)
        {
            foreach(Product product in products)
            {
                if (product.Size == size) yield return product;
            }
        }

        public IEnumerable<Product> FilterBySizeAndColor(IEnumerable<Product> products, Size size, Color color)
        {
            foreach (Product product in products)
            {
                if (product.Color == color && product.Size == size) yield return product;
            }
        }
    }

    // Create 2 new Interfaces that are Open for Extension
    public interface ISpecification<T>
    {
        bool IsSatisfied(T item);
    }
    public interface IFilter<T>
    {
        IEnumerable<T> Filter(IEnumerable<T> items, ISpecification<T> spec);
    }

    public class SizeSpecification : ISpecification<Product>
    {
        public Size Size { get; set; }

        public SizeSpecification(Size size)
        {
            Size = size;
        }

        public bool IsSatisfied(Product product)
        {
            return product.Size == Size;
        }
    }

    public class ColorSpecification : ISpecification<Product>
    {
        public Color Color { get; set; }

        public ColorSpecification(Color color)
        {
            Color = color;
        }

        public bool IsSatisfied(Product product)
        {
            return product.Color == Color;
        }
    }

    // Combinator
    public class AndSpecification<T> : ISpecification<T>
    {
        public ISpecification<T> First { get; set; }
        public ISpecification<T> Second { get; set; }
        public AndSpecification(ISpecification<T> first, ISpecification<T> second)
        {
            First = first ?? throw new ArgumentNullException(paramName: nameof(first));
            Second = second ?? throw new ArgumentNullException(paramName: nameof(second));
        }

        public bool IsSatisfied(T item)
        {
            return First.IsSatisfied(item) && Second.IsSatisfied(item);
        }
    }

    public class BetterFilter : IFilter<Product>
    {
        public IEnumerable<Product> Filter(IEnumerable<Product> items, ISpecification<Product> spec)
        {
            foreach (var item in items)
            {
                if (spec.IsSatisfied(item)) yield return item;
            }
        }
    }
}
